### DISCUSSION TITLE
Limiting Abuse Control (rate-limiting) to Client-Side Integrations

### DISCUSSION THREAD

**Discussion by eldadfux** [appwrite/appwrite#813] [Ideas]
I'm thinking of limiting the abuse checks to client-side integrations only for all Appwrite different endpoints. I think it's becoming a bit limiting and counterproductive when using Appwrite with an API key.

We use rate-limiting today for mainly two scenarios. The first one is the protection of user authentication, and the second one is protecting write-heavy operations. Both cases are much more dangerous when we're talking about end-users access rather than just an internal server integration using an API key or upcoming JWT authentication methods.

I would love to get more community and team feedback about this before making any changes in the code.

**Comment by punitdiwan** [comment]
I think you can keep Abuse Control on Client Side and Server side as well, but keep it configurable using .env Variables with default values and document the same. It will make both client and server users happy.

**Comment by silentworks** [comment]
I think limiting it to client-side only is fine, but it would be good if the rate limit could be changed by the person setting up the Appwrite server. I don't like the idea of a hardcoded limiting set by the creators of the software themselves and not being configurable.

**Reply by eldadfux** [reply]
Do you think we should make this configurable at project or server level?

**Reply by kodumbeats** [reply]
Since most of Appwrite dev/user experience is handled at the project level, I think API rate limiting per project is appropriate.

**Comment by TorstenDittmann** [comment]
In my opinion, there should be no rate limit in conjunction with an API key.

API keys of this type are sensitive information and are usually treated as such, so there are completely different problems with a leaked key. Also, there are use cases in the server side - where a rate limit might actually compromise functionality. 

The suggestion from @silentworks also sounds like a very good idea! 👍🏻

**Reply by eldadfux** [reply]
I 100% agree that this can actually compromise functionality on the server side.

**Comment by lohanidamodar** [comment]
I agree with @TorstenDittmann, Rate-Limit on requests with API keys doesn't make much sense. But in client side its a great security addition. But, yes it should be configurable, and best to have it configurable per project. As, different projects might have different requirement and sensitivity.

**Comment by stillie** [comment]
Apologies for the hijack, delete if need be, but i was wondering is rate limiting configurable? If so where? If not, is there plans to implement it?

**Reply by stnguyen90** [reply]
The only rate limit configuration is to enable/disable. See [docs](https://appwrite.io/docs/rate-limits#disable-limits).

**Comment by stnguyen90** [comment]
It typically makes sense to only impose the abuse controls client side, but what about when Appwrite goes cloud/multi-tenant? What controls will be in place to prevent one org's project from using a server side api key to impact the performance of other projects and the entire server?

**Reply by eldadfux** [reply]
We have plans to handle this kind of miss-usage in multiple levels. Since this is sensitive security information about the Cloud service that could be used for abuse attempts, we better not disclose it publicly.

**Comment by brianmutiso** [comment]
I am here because I started using Appwrite yesterday, Appwrite is such a wonderful product,
Rate limiting on the client side can affect the developer, for example, right now I have to wait for 60 mins to continue development and testing, I think it should be configurable per project

**Reply by brianmutiso** [reply]
Does Rate limiting affect several devices on the same IP? for instance when an app is used by several people on the same public IP some will find it unusable, so allowing configuration per project might really help for such scenarios

**Reply by gewenyu99** [reply]
:wave: We suggest disabling rate limiting during development. You can do this through changing the variable `_APP_OPTIONS_ABUSE=disabled`. Remember to enable it again once you're in a production environment!

https://appwrite.io/docs/environment-variables

**Comment by Shadowfita** [comment]
I completely agree with the idea of removing rate limits from the projects API key. I don't think it encourages large scale functionality by having that limit in place on a project level.

Limit users / client endpoints is a must though.

**Comment by brianmutiso** [comment]
Yes, i managed to disable it

**Comment by ghost** [comment]
I think it should be configurable on both server and client too.

I've run into a problem where I can't use Appwrite since it is likely many people will create accounts at the same time.

**Reply by TorstenDittmann** [reply]
The rate limits for this also take the users IP into account. So the limits are not globally applied.

**Reply by ghost** [reply]
@TorstenDittmann

The user's local IP on their network? Or their network's public facing IP?

I should have mentioned that. It's likely there will be multiple accounts created at the same same time on the same network (like in a classroom)

**Comment by devberkay** [comment]
For example , i want to restrict usage of "Password Recovery" functionality of Account API to 2 requests maximum per 60 minutes but official documents says "This endpoint is limited to 10 requests in every 60 minutes per email address and IP address" , is it possible to make this happen?

**Reply by stnguyen90** [reply]
It's not configurable right now. You would have to modify this line before the appwrite container starts: https://github.com/appwrite/appwrite/blob/e3c9041ca1c8fdb374334c06e54d95d50a59da44/app/controllers/api/account.php#L1851.

**Comment by stnguyen90** [comment]
As of [Version 0.7.0](https://github.com/appwrite/appwrite/releases/tag/0.7.0), [rate limits](https://appwrite.io/docs/rate-limits) only apply to clients and are ignored when accessing Appwrite with an API key.
